File: app/src/main/java/io/oldering/tvfoot/red/match/state/MatchResult.java
package io.oldering.tvfoot.red.match.state;

import com.google.auto.value.AutoValue;
import io.oldering.tvfoot.red.data.entity.Match;
import javax.annotation.Nullable;

import static io.oldering.tvfoot.red.match.state.MatchResult.Status.LOAD_MATCH_FAILURE;
import static io.oldering.tvfoot.red.match.state.MatchResult.Status.LOAD_MATCH_IN_FLIGHT;
import static io.oldering.tvfoot.red.match.state.MatchResult.Status.LOAD_MATCH_SUCCESS;

interface MatchResult {
  @AutoValue abstract class LoadMatchResult implements MatchResult {
    abstract Status status();

    @Nullable abstract Match match();

    @Nullable abstract Throwable error();

    static LoadMatchResult success(Match match) {
      return new AutoValue_MatchResult_LoadMatchResult(LOAD_MATCH_SUCCESS, match, null);
    }

    static LoadMatchResult failure(Throwable throwable) {
      return new AutoValue_MatchResult_LoadMatchResult(LOAD_MATCH_FAILURE, null, throwable);
    }

    static LoadMatchResult inFlight() {
      return new AutoValue_MatchResult_LoadMatchResult(LOAD_MATCH_IN_FLIGHT, null, null);
    }
  }

  @AutoValue abstract class GetLastStateResult implements MatchResult {
    static GetLastStateResult create() {
      return new AutoValue_MatchResult_GetLastStateResult();
    }
  }

  enum Status {
    LOAD_MATCH_IN_FLIGHT, LOAD_MATCH_FAILURE, LOAD_MATCH_SUCCESS, //
  }
}

File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/match/MatchResult.kt
package com.benoitquenaudon.tvfoot.red.app.domain.match

import com.benoitquenaudon.tvfoot.red.app.data.entity.Match
import com.benoitquenaudon.tvfoot.red.app.mvi.MviResult

sealed class MatchResult : MviResult {
  sealed class LoadMatchResult : MatchResult() {
    data class Success(val match: Match, val shouldNotifyMatchStart: Boolean) : LoadMatchResult()

    data class Failure(val throwable: Throwable) : LoadMatchResult()

    object InFlight : LoadMatchResult()
  }

  data class NotifyMatchStartResult(val shouldNotifyMatchStart: Boolean) : MatchResult()
}
